reactioncommerce/reaction
# (feat) Use token to create Stripe charges

Resolves #4323 
Impact: major
Type: feature

## Issue
Sensitive card data touch our server prior to submitting to Stripe which is not the ideal security practice.

## Solution
Tokenize credit card data on the client using [Stripe Elements](https://stripe.com/elements) which are pre-built Stripe UI components. This way, only a token will reach the backend that can be used to create charges. The source code for the React components is [here](https://github.com/stripe/react-stripe-elements).

## Breaking changes
1. The checkout screen will change.
- Name is no longer required.
- The card number input creates a space every 4 characters.
- Expiry month and year are in a single input, no dropdown for month selection.
- Postal code is required.
- There is field validation but errors are not linked to a field, so all error messages are displayed just above the submit button and are determined only after user submits the form.
- Spacing and font are not yet exactly the same as our other payment method implementations but I believe these are relatively easier to fix as compared to the other mentioned changes.

<img width="338" alt="screen shot 2561-06-14 at 17 29 53" src="https://user-images.githubusercontent.com/10388757/41407101-9ff5c732-6ff8-11e8-8bec-9345d3b431a3.png">

2. Added a public field `settings.public.publishable_key` in setting up the primary shop's Stripe credentials. It is used the public key to tokenize credit card data on the client.

<img width="411" alt="screen shot 2561-06-14 at 17 16 05" src="https://user-images.githubusercontent.com/10388757/41406497-bf4edecc-6ff6-11e8-8e66-1630684a0bd9.png">

3. Added `react-stripe-elements` as a dependency.
4. Revised the test for `stripe/payment/createCharges`. This PR modifies the function mentioned to accept a `token` object instead of `cardData`. Hence, there is a necessity to revise the test to adjust the parameters needed for it to pass. The expected results from the old test was not changed, so there should not be a huge impact.

Aside from this test, I added 2 additional tests for `stripe/payment/createCharges` for better coverage namely:
`should throw an error with an invalid api key` and
`should be able to create multiple charges`

## Testing
1. Set up Stripe credentials in the Payment Dashboard (Secret API Key, Publishable Key, Public Client ID and percentage fee.)
2. Create an order and pay.
3. As an admin, you should be able to capture and refund the payment.

To test marketplace payments:
1. Turn on the marketplace feature. Invite a shop owner and activate the shop.
2. Log in as a shop owner and create and publish a product.
3. In Shop Settings Dashboard, update the shop address and click Start Accepting Payment under Setup Stripe Account.
4. As a guest, buy items from the primary shop and the new shop.
5. After payment, the admin and shop owner should be able to capture and refund payment of customer.
